### PR TITLE
SMP: add loading placeholder

### LINKED ISSUE #66203: Sites Management Page: Loading state while fetching sites
This is what the loading state looks like currently.

Image

It's indistinguishable from the case where users actually have no sites.

The usual pattern in Calypso is to show a placeholder that roughly resembles what the shape of what the loaded state looks like, except with pulsating rectangles. See this standard component for an idea: https://wpcalypso.wordpress.com/devdocs/blocks/site-placeholder. Take a look at the mixin in the `_placeholder.scss` for implementing the animation.

A possible implementation might be a simplified table with two or three rows that shows rectangles in place of the text and site icon.

Done is:
- The loading state for the sites table resembles a loaded sites table
- The pulsating animation should look the same as it does elsewhere in Calypso (use the mixin)
- If we've implemented a tile view
   - The loading state for tile view resembles a loaded tile view

### PR DESCRIPTION
**Pull request by sejas**
#### Proposed Changes

* Create new package/component `LoadingPlaceholder`
* Add component stories to storybook
* Add loading placeholders while fetching for the sites
* Add animation delay to the rows to improve the design

66203_dynamic-rows-delay

#### Screencast

https://user-images.githubusercontent.com/779993/182966595-9ff3645f-2f28-4a92-ba81-ba38f5c6103a.mp4

#### Testing Instructions

* Go to `/sites-dashboard`
* Open Chrome Devtools > Application > IndexedDB > and delete it
*  Refresh the page
* Observe the loading placeholders are visible for a brief moment until the sites are loaded.

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Closes #66203

### CODE REVIEW

File: client/data/sites/use-site-excerpts-query.ts
Comment (vykes-mac) [inline client/data/sites/use-site-excerpts-query.ts]: just a little typo here, nothing major

**Review by vykes-mac — COMMENTED** [review]
This looks good @sejas , eventually we want to resize the LoadingLogo styled component to reflect the size and radius of the site thumbnail.

File: client/data/sites/use-site-excerpts-query.ts
Comment (p-jackson) [inline client/data/sites/use-site-excerpts-query.ts]: [Reading the docs for `isLoading`](https://react-query-v3.tanstack.com/reference/useQuery) it seems like it's exactly what we want, and I think it's best to leave the meaning for these standard ReactQuery objects as-is so that other devs who are familiar with ReactQuery know how they work.

I think the issue is the `placeholderData` option we pass to `useQuery`. Since we're providing a placeholder it means the query is never in a "hard" loading state, and so `isLoading` is always false. Now that you're implementing a proper placeholder UI I think we should remove `placeholderData` option and leave the `isLoading` flag as-is.

File: client/sites-dashboard/components/sites-table.tsx
Comment (p-jackson) [inline client/sites-dashboard/components/sites-table.tsx]: An alternative would be to have each component know how to render an `isLoading` version of itself i.e.

```suggestion
						<SitesTableRow isLoading={ true } />
						<SitesTableRow isLoading={ true } />
						<SitesTableRow isLoading={ true } />
```

The row would render a `isLoading` version of itself and then pass the prop down to its sub-components. That way if we add another column we don't have to update both `<SitesTableRow>` _and_ `<SitesTableRowLoading>`. And as @vykes-mac pointed out, the thumbnail component will be responsible for knowing how big to render the placeholder version of itself. As it is we're already having to duplicate the `mobileHidden` code.

The downside is that `site` would become an optional prop and so there'd have to be a bunch of `undefined` handling inside `<SitesTableRow>` which is currently nice a clean from that sort of thing.

File: client/sites-dashboard/components/sites-table-row-loading.tsx
Comment (p-jackson) [inline client/sites-dashboard/components/sites-table-row-loading.tsx]: It's too bad we can't reuse the `placeholder()` SCSS mixin here in Emotion, so that we resuse the same animation as the rest of Calypso.
https://github.com/Automattic/wp-calypso/blob/7e0a33b4ac44dae1aca27d2d403bbb9a2a476431/client/assets/stylesheets/shared/mixins/_placeholder.scss#L6

We could add a SCSS file to the project just for this one thing. Up to you, it's probably fine to have our own version (I bet there's already a lot of duplicates of this animation in Calypso).

**Review by p-jackson — DISMISSED** [review]
Feels much more satisfying having this in place!
The only thing I think we definitely should change is the `isLoading` property from `useSiteExcerptsQuery()`.

File: client/data/sites/use-site-excerpts-query.ts
Comment (sejas) [reply client/data/sites/use-site-excerpts-query.ts]: @p-jackson Thank you!! , Yesterday I tried to remove the `initialData` to solve the isLoading issue, but I didn't try removing the `placeholderData`.

Now the default `isLoading` is working as expected 💪

File: client/sites-dashboard/components/sites-table.tsx
Comment (sejas) [reply client/sites-dashboard/components/sites-table.tsx]: I've decided to keep the `SitesTableRowLoading` separate from `SitesTableRow`, but now we don't need to update `SitesTableRowLoading` to add/remove the number of columns. We just need to change the prop `columns={5}` https://github.com/Automattic/wp-calypso/blob/172c2ed745b7ad9c4a986154494c0104c8956fdb/client/sites-dashboard/components/sites-table.tsx#L56-L64

File: client/sites-dashboard/components/sites-table-row-loading.tsx
Comment (sejas) [reply client/sites-dashboard/components/sites-table-row-loading.tsx]: I've created a new component, `LoadingPlaceholder` inside packages. Now all the logic and styles live in that component with new enhancements such as adding an animation delay.
I've also added new examples in the storybook.

Screenshot 2022-08-05 at 17 47 36

With this approach, our components will be more reusable in different applications.

https://github.com/Automattic/wp-calypso/blob/172c2ed745b7ad9c4a986154494c0104c8956fdb/packages/components/src/loading-placeholder/loading-placeholder.tsx#L1-L21

Comment (sejas) [reply client/sites-dashboard/components/sites-table-row-loading.tsx]: I've created a new component, `LoadingPlaceholder` inside packages. Now all the logic and styles live in that component with new enhancements such as adding an animation delay.
I've also added new examples in the storybook.

Screenshot 2022-08-05 at 17 47 36

With this approach, our components will be more reusable in different applications.

Using the placeholder mixin it works on calypso, but not in the storybook.
This is the error I get in the storybook using the mixin:
```
index.js:56 Unexpected error while loading ./loading-placeholder/loading-placeholder.stories.tsx: Module build failed (from ../../node_modules/sass-loader/dist/cjs.js):
SassError: Undefined mixin.
  ╷
2 │     @include placeholder();
  │     ^^^^^^^^^^^^^^^^^^^^^^
  ╵
  src/loading-placeholder/style.scss 2:2  root stylesheet
 Error: Module build failed (from ../../node_modules/sass-loader/dist/cjs.js):
```

https://github.com/Automattic/wp-calypso/blob/172c2ed745b7ad9c4a986154494c0104c8956fdb/packages/components/src/loading-placeholder/loading-placeholder.tsx#L1-L21

**Comment by sejas** [comment]
I've added a small delay animation between rows:

**OLD SAME TIME**

66203_loading-placeholder

**WITH DELAY**

66203_dynamic-rows-delay

**Comment by sejas** [comment]
@p-jackson @vykes-mac Could you take another look? I've refactored the code and created a new package component, `LoadingPlaceholder`. This component can be used outside of calypso.

**Comment by vykes-mac** [comment]
@sejas I wonder if this component should be added to the `/devdocs/design` here client/devdocs/design/index.jsx? I'm not certain when we add it to the devdocs vs the storybook, storybook does seem legacy though.

**Review by vykes-mac — COMMENTED** [review]
This looks good! I really like the delay between the rows make the animation more satisfying.

File: client/sites-dashboard/components/sites-dashboard.tsx
Comment (zaguiini) [inline client/sites-dashboard/components/sites-dashboard.tsx:97]: I don't really like having to pass the `isLoading` prop to the `SitesTable` component. It feels like we're delegating the decision that affects the behavior of the whole page to an inner component. Here's how I would do it:

1. I would change `SitesTable` to accept a `children` prop, which would be the content of the table, like `<tbody>{children}</tbody>`;
2. In this `SiteDashboard` component, I would modify the expression to something like:

```tsx
{ filteredSites.length > 0 ? <SitesTable>{ sites.map(...) }</SitesTable> :
isLoading ? <SitesTable><SitesTablePlaceholder /></SitesTable> :
<NoSitesMessage {...} /> }
```

3. `<SitesTablePlaceholder />` would be created and would hold:

```tsx
Array( N_LOADING_ROWS )
						.fill( null )
						.map( ( _, i ) => (
							<SitesTableRowLoading
								columns={ 5 }
								delayMS={ i * 150 }
								logoProps={ { width: 108, height: 78 } }
							/>
						) )
```

Let me know your thoughts on this approach. Another possibility would be to remove the conditionals that prevent the rendering of `<SitesTable />` and then making the decision to display it or not inside the component itself. That's more obscure to me and what's happening is not immediately visible, so I don't like this approach that much...

File: client/sites-dashboard/components/sites-table.tsx
Comment (p-jackson) [inline client/sites-dashboard/components/sites-table.tsx:59]: This is generating a React warning about a missing key. One of the few times I think it's ok to just use the array index as the key I think :)

```suggestion
							<SitesTableRowLoading
								key={ i }
```

File: packages/components/src/index.js
Comment (p-jackson) [inline packages/components/src/index.js]: Build error occurring here. You can't export a type from a `*.js` file.

File: packages/components/src/loading-placeholder/loading-placeholder.tsx
Comment (p-jackson) [inline packages/components/src/loading-placeholder/loading-placeholder.tsx]: Is this a typo?

```suggestion
const pulseLightKeyframes = keyframes`
```

File: client/sites-dashboard/components/sites-dashboard.tsx
Comment (p-jackson) [reply client/sites-dashboard/components/sites-dashboard.tsx:97]: > It feels like we're delegating the decision that affects the behavior of the whole page to an inner component.

The `isLoading` prop seems good to me. The parent component keeps ownership of the _decision_ about whether to render the loading state, but it's delegating the visual details to the table component. For instance what if the table headings also needed a loading state? (e.g. maybe the number of columns depends on the site data) We'd have to refactor `<SitesDashboard>`, having it switch between `<SitesTableHeadings>` and `<SitesTableHeadingsPlaceholder>`.

I do agree that using flags to switch between modes multiple modes for a component can get out of hand! Some of the components in the signup code have lots of modes and it is pretty confusing. But I think a flag for "render a loading version of yourself" can be generally understood in Calypso because it's a design pattern we use throughout the app.

**Review by p-jackson — COMMENTED** [review]
The delay looks cool 😎 
Thanks for the changes. The build is failing because the `LoadingPlaceholderProps` type is being exported from a `*.js` file. There are three solutions I can think of:
1. Rename `packages/components/src/index.js` to `packages/components/src/index.ts`
2. Import the type directly from its source file i.e.
	`import type { LoadingPlaceholderProps } from '@automattic/components/src/loading-placeholder';`
3. Don't worry about extending the `LoadingPlaceholderProps` interface and just add the `delayMS` prop directly to `SitesTableRowLoadingProps`. If the props ever get out of sync down the track there'll be a TypeScript error to tell us about it.

I think I lean towards (3).

File: client/sites-dashboard/components/sites-table.tsx
Comment (sejas) [reply client/sites-dashboard/components/sites-table.tsx:59]: Nice catch! I didn't check the console, and I always forget the key.

**Comment by sejas** [comment]
@Automattic/build I've fixed the build errors and other improvements.
Waiting for final approval.

**Comment by sejas** [comment]
> @sejas I wonder if this component should be added to the `/devdocs/design` here client/devdocs/design/index.jsx? I'm not certain when we add it to the devdocs vs the storybook, storybook does seem legacy though.

@vykes-mac , I think we should try to create components that can be reused in different applications and not only in calypso. For this purpose, we can create the components inside packages/src, and the best way to develop isolated components is using storybook. You can develop the component faster thanks to the default react hot reloading, and you can make sure your component will work outside of calypso.

File: client/sites-dashboard/components/sites-table-row-loading.tsx
Comment (zaguiini) [inline client/sites-dashboard/components/sites-table-row-loading.tsx:53]: Let's use CSS objects instead of style declaration soup 😛

File: client/sites-dashboard/components/sites-table-row-loading.tsx
Comment (zaguiini) [inline client/sites-dashboard/components/sites-table-row-loading.tsx]: Instead of defining the default value here, we could directly put the props before `{ ...logoProps }` and that'd bring us exactly the same output. The difference is that if people want to pass additional `logoProps` that are not `width` and `height`, they'll lose the default width and height set here.

File: packages/components/src/loading-placeholder/loading-placeholder.tsx
Comment (zaguiini) [inline packages/components/src/loading-placeholder/loading-placeholder.tsx:19]: Let's prefer CSS objects. TS inference FTW!

**Review by zaguiini — COMMENTED** [review]
Almost there 💪

File: client/sites-dashboard/components/sites-table-row-loading.tsx
Comment (danielbachhuber) [reply client/sites-dashboard/components/sites-table-row-loading.tsx]: @zaguiini Great suggestion! Updated with 691a025299

File: client/sites-dashboard/components/sites-table-row-loading.tsx
Comment (danielbachhuber) [reply client/sites-dashboard/components/sites-table-row-loading.tsx:53]: > Let's use CSS objects instead of style declaration soup 😛

@zaguiini Can you provide an example? @sejas' implementation seems quite common across the codebase. Not sure which alternative pattern you're referring to.

**Comment by danielbachhuber** [comment]
Great work on the placeholders, @sejas ! It seems like the right amount of abstraction.

File: client/sites-dashboard/components/sites-table-row-loading.tsx
Comment (zaguiini) [reply client/sites-dashboard/components/sites-table-row-loading.tsx:53]:
```js
const MyStyledComponent = styled.element({
   rule: 'value'
})

**Pull request merged**
